src/Control/Monad/Catch.hs

```diff
@@ -733,72 +747,93 @@ instance MonadCatch m => MonadCatch (ListT m) where
 ------------------------------------------------------------------------------
 
 -- | Like 'mask', but does not pass a @restore@ action to the argument.
-mask_ :: MonadMask m => m a -> m a
+mask_ :: (HasCallStack, MonadMask m) => m a -> m a
 mask_ io = mask $ \_ -> io
 
 -- | Like 'uninterruptibleMask', but does not pass a @restore@ action to the
 -- argument.
-uninterruptibleMask_ :: MonadMask m => m a -> m a
-uninterruptibleMask_ io = uninterruptibleMask $ \_ -> io
+uninterruptibleMask_ :: (HasCallStack, MonadMask m) => m a -> m a
+uninterruptibleMask_ io = uninterruptibleMask (\_ -> io)
+
+-- You may wonder why there isn't a `withFrozenCallStack` on `mask_` and
+-- `uninterruptibleMask_`. The reason is that GHC is incapable of unifying
+-- the type due to the RankNType in `mask`.
+--
+-- src/Control/Monad/Catch.hs:746:32: error:
+--     • Couldn't match type ‘p1’ with ‘forall a1. m a1 -> m a1’
+--       Expected: (forall a1. m a1 -> m a1) -> m a
+--         Actual: p1 -> m a
+--       Cannot instantiate unification variable ‘p1’
+--       with a type involving polytypes: forall a1. m a1 -> m a1
+--     • In the first argument of ‘withFrozenCallStack’, namely ‘mask’
+--       In the first argument of ‘($)’, namely ‘withFrozenCallStack mask’
+--       In the expression: withFrozenCallStack mask $ \ _ -> io
+--     • Relevant bindings include
+--         io :: m a (bound at src/Control/Monad/Catch.hs:746:7)
+--         mask_ :: m a -> m a (bound at src/Control/Monad/Catch.hs:746:1)
+--     |
+-- 746 | mask_ io = withFrozenCallStack mask $ \_ -> io
+--     |                                ^^^^
```

Comment:
I'm confused. Why not just write this?

```hs
 mask_ io = withFrozenCallStack (mask (\_ -> io))
```

Alternatively, if you really want to only apply `withFrozenCallStack` to just `mask`, you can manually eta expand to get:

```hs
mask_ io = withFrozenCallStack (\f -> mask (\x -> f x)) (\_ -> io)
```

This will work even with simplified subsumption in GHC 9.0+. But then again, I'm 99% that these will produce the same code in the end, so you might as well pick the former option, which is simpler.